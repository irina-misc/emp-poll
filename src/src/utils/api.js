  import {
    _getUsers,
    _getQuestions
  } from "./_DATA.js";


  export function getInitialData () {
    return Promise.all([
      getUsers(),
      getQuestions(),
    ]).then(([users, questions]) => ({
      users,
      questions,
    }))
  }

  export function getUsers() {
    return _getUsers();    
  }

  export function getQuestions() {
    return _getQuestions();
  }


