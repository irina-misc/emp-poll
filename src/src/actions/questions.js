import { showLoading, hideLoading } from "react-redux-loading-bar";
import {_saveQuestion ,_saveQuestionAnswer} from "../utils/_DATA";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ANSWER_QUESTION = "ANSWER_QUESTION";
export const ADD_QUESTION = "ADD_QUESTION";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

function answerQuestion({ authedUser, qid, answer }) {
  return {
    type: ANSWER_QUESTION,
    qid,
    authedUser,
    answer,
  };
}

export function handleAddQuestion(question) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
     question["author"] = authedUser;
    dispatch(showLoading());
    return _saveQuestion(
      question,
    )
      .then((question) => dispatch(addQuestion(question)))
      .then(() => dispatch(hideLoading()));
  };
}

export function handleAnswerQuestion({ authedUser, qid, answer }) {
  return (dispatch) => {
    dispatch(answerQuestion({ authedUser, qid, answer }))
    dispatch(showLoading());
    return _saveQuestionAnswer({ authedUser, qid, answer }).then(
      dispatch(hideLoading())
    )
     .catch((e) => {
       console.warn("Error in handleAnswerQuestion: ", e);
       alert("The was an error liking the answer. Try again.");
     }).then;
  };
}
