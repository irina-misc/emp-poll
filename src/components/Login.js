
import { connect } from "react-redux";
import { useState } from "react";
import { handleSetAuthedUser } from "../actions/authedUser"

const Login = (props) => {
    const [selectedUser, setSelectedUser] = useState(null);

    if (props.authedUser || props.authedUser !== null) {
        return null;
    }

    const handleChange = (e) => {
        console.log("e.target.value=", e.target.value);
        setSelectedUser(e.target.value);
    };

    const handleLogon = (e) => {
        props.dispatch(handleSetAuthedUser(selectedUser));
    }

    return (<div className="card">
        <div className="card-header" data-testid="login-card">
            <h3>Log On to Employee Polls</h3>
        </div>

        <div className="card-body">
            <form data-testid="log-form" onSubmit={(e) => handleLogon(e)}>
                <div>
                    <select data-testid="select-user" role="combobox" style={{ width: "15.5em", marginBottom: 30 }} value={selectedUser !== null ? selectedUser : 'n/a'} onChange={handleChange}>
                        <option role="option" key="1" value="n/a" disabled >Select user...</option>
                        {props.users && Object.keys(props.users).map(id => {
                            return (<option key={id} value={id} role="option">{props.users[id].name}</option>)
                        })}
                    </select>
                </div>
                <button className="btn btn-secondary" type="submit"
                    disabled={selectedUser === null}
                    data-testid="log-button">
                    Log On
                </button>
            </form>
        </div>
    </div>
    )
};

function mapStateToProps({ users, authedUser }) {
    return {
        users: users,
        authedUser: authedUser
    };
}

export default connect(mapStateToProps)(Login);
