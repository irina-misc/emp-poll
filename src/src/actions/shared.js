
import { getUsers, getQuestions }from "../utils/api";
import { receiveUsers } from "./users";
import { receiveQuestions } from "./questions";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export  function handleInitialData() {
  return (dispatch) => {
    dispatch(showLoading());
    return Promise.all([getUsers(), getQuestions()]).then(
      ([users, question]) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(question));
        dispatch(hideLoading());
      }
    );
  };
}