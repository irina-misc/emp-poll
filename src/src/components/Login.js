
import { connect } from "react-redux";
import { useState } from "react";
import { handleSetAuthedUser } from "../actions/authedUser"

const Login = (props) => {
    const [selectedUser, setSelectedUser] = useState(null);

    if (props.authedUser || props.authedUser !== null) {
        return null;
    }

    const handleChange = (e) => {
        console.log("e.target.value=", e.target.value);
        setSelectedUser(e.target.value);
    };

    const handleLogon = (e) => {
        props.dispatch(handleSetAuthedUser(selectedUser));
    }

    return (<div className="card">
        <div className="card-header" data-testid="login-card">
            <h3>Log On to Employee Polls</h3>
        </div>
        <div className="card-body">
            <form onSubmit={(e) => handleLogon(e)}>
                <div>
                    <select style={{ width: "15.5em", marginBottom: 30 }} value={selectedUser !== null ? selectedUser : 'n/a'} onChange={handleChange}>
                        <option key="1" value="n/a" disabled >Select user...</option>
                        {props.users && Object.keys(props.users).map(id => {
                            return (<option key={id} value={id} >{props.users[id].name}</option>)
                        })}
                    </select>
                </div>
                <button className="btn btn-secondary" type="submit"
                    disabled={selectedUser === null}
                    data-testid="Logon">
                    Log On
                </button>
            </form>
        </div>
    </div>
    )
};

function mapStateToProps({ users, authedUser }) {
    return {
        users: users,
        authedUser: authedUser
    };
}

export default connect(mapStateToProps)(Login);
