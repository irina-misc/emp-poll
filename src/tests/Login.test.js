import * as React from 'react';
import { createStore } from "redux";
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from "react-redux";
import reducer from "../reducers";
import middleware from "../middleware";
import '@testing-library/jest-dom'
import { MemoryRouter } from 'react-router';
import { handleInitialData } from "../actions/shared"
import Login from '../components/Login';
import {getUsers} from '../utils/api';

describe("Login", () => {
  
  it("Login component render snapshot", () => {
    const store = createStore(reducer, middleware);
    store.dispatch(handleInitialData());
    const component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it("Log On button should be disabled - no user selected", async () => {
    const store = createStore(reducer, middleware);
    const  users = getUsers();
    
    const component = render(
      <Provider store={store}>
        <Login users={users} />
      </Provider>
    );
    var logForm = component.getByTestId('log-form')
    var logOnButton = component.getByTestId('log-button')
    expect(logForm).toBeInTheDocument();
    expect(logOnButton).toBeDisabled();
  });


  it('Log On button should be enabled - user selected', () => {
    const store = createStore(reducer, middleware);
    const  users = getUsers();
    
    const component = render(
      <Provider store={store}>
        <Login users={users} />
      </Provider>
    );

    var select = component.getByTestId('select-user');
    fireEvent.change(select, { target: { value: 'sarahedo' } });

    var logForm = component.getByTestId('log-form')
    var logOnButton = component.getByTestId('log-button');
    expect(logForm).toBeInTheDocument();
    expect(logOnButton).toBeEnabled();

  })
});