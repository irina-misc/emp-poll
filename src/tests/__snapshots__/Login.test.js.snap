// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login component render snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="card-header"
          data-testid="login-card"
        >
          <h3>
            Log On to Employee Polls
          </h3>
        </div>
        <div
          class="card-body"
        >
          <form
            data-testid="log-form"
          >
            <div>
              <select
                data-testid="select-user"
                style="width: 15.5em; margin-bottom: 30px;"
              >
                <option
                  disabled=""
                  value="n/a"
                >
                  Select user...
                </option>
              </select>
            </div>
            <button
              class="btn btn-secondary"
              data-testid="log-button"
              disabled=""
              type="submit"
            >
              Log On
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
    >
      <div
        class="card-header"
        data-testid="login-card"
      >
        <h3>
          Log On to Employee Polls
        </h3>
      </div>
      <div
        class="card-body"
      >
        <form
          data-testid="log-form"
        >
          <div>
            <select
              data-testid="select-user"
              style="width: 15.5em; margin-bottom: 30px;"
            >
              <option
                disabled=""
                value="n/a"
              >
                Select user...
              </option>
            </select>
          </div>
          <button
            class="btn btn-secondary"
            data-testid="log-button"
            disabled=""
            type="submit"
          >
            Log On
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
